File: Beecrowd_Parte2/Elementar_Caro_Watson_1382.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void * a, const void * b) {
    return ( *(int*)a - *(int*)b );
}

int minSwapsToSort(int arr[], int N) {
    int temp[N];
    for (int i=0; i<N; i++)
        temp[i] = arr[i];

    qsort(temp, N, sizeof(int), compare);

    int visited[N];
    for (int i=0; i<N; i++)
        visited[i] = 0;

    int swaps = 0;

    for (int i = 0; i < N; i++) {
        if (visited[i] || temp[i] == arr[i])
            continue;

        int cycle_size = 0;
        int j = i;

        while (!visited[j]) {
            visited[j] = 1;
            j = find(temp, N, arr[j]);
            cycle_size++;
        }

        if (cycle_size > 0) {
            swaps += (cycle_size - 1);
        }
    }

    return swaps;
}

int find(int arr[], int n, int target) {
    for (int i = 0; i < n; i++)
        if (arr[i] == target)
            return i;
    return -1;
}

int main() {
    int T;
    scanf("%d", &T);

    for (int t = 0; t < T; t++) {
        int N;
        scanf("%d", &N);

        int arr[N];
        for (int i = 0; i < N; i++) {
            scanf("%d", &arr[i]);
        }

        printf("%d\n", minSwapsToSort(arr, N));
    }

    return 0;
}

File: Beecrowd_Parte2/Elementar_Caro_Watson_1382.java
import java.util.*;

public class Elementar_Caro_Watson_1382 {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();

        for (int t = 0; t < T; t++) {
            int N = sc.nextInt();
            int[] arr = new int[N];

            for (int i = 0; i < N; i++) {
                arr[i] = sc.nextInt();
            }

            System.out.println(minSwapsToSort(arr, N));
        }

        sc.close();
    }

    public static int minSwapsToSort(int[] arr, int N) {
        int[] temp = Arrays.copyOf(arr, N);
        Arrays.sort(temp);
        Map<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < N; i++) {
            map.put(temp[i], i);
        }

        boolean[] visited = new boolean[N];
        Arrays.fill(visited, false);

        int swaps = 0;

        for (int i = 0; i < N; i++) {
            if (visited[i] || map.get(arr[i]) == i) {
                continue;
            }

            int cycleSize = 0;
            int j = i;

            while (!visited[j]) {
                visited[j] = true;
                j = map.get(arr[j]);
                cycleSize++;
            }

            if (cycleSize > 0) {
                swaps += (cycleSize - 1);
            }
        }

        return swaps;
    }
}
